intra_process_demo/README.md

```diff
@@ -0,0 +1,171 @@
+## **What Is This?**
+
+This demo is designed to showcase how developers can **manually compose ROS 2 nodes** by **defining them separately** but **combining them in varied process layouts**.
+
+These can be done while avoiding code overhauls or performance restrictions. 
+
+This package consists of the following ROS 2 nodes:
+
+1. `camera_node`
+2. `watermark_node`
+3. `image_view_node`
+4. `image_pipeline_all_in_one`
+5. `two_node_pipeline`
+6. `cyclic_pipeline`
+7. `image_pipeline_with_two_image_view`
+
+Through the use of **intra-process** (as opposed to **inter-process** node communication), lower latency and thus **higher efficiency** is observed for ROS 2 topologies that utilizes this manner of communication. 
+
+These improvements in latency reduction is even **more pronounced when applied to ROS 2 systems with more complex topology**. 
+
+## **Build**
+
+```bash
+colcon build --packages-up-to intra_process_demo
+```
+
+## **Run**
+
+### 1. Two Node Pipeline
+
+Run `two_node_pipeline` via the commands below:
+
+```bash
+ros2 run intra_process_demo two_node_pipeline
+```
+
+> This sets up two nodes, a ROS 2 node that publishes a string with an incremeting integer value, as well as a ROS 2 node that subscribes to print the received string.
+
+![](img/two_node_pipeline.png)
+
+### 2. Cyclic Pipeline
+
+Run `cyclic_pipeline` via the commands below:
+
+```bash
+ros2 run intra_process_demo cyclic_pipeline
+```
+
+> Similar to the previous, instead of creating a new message for each new iteration, the publisher and subscriber nodes only ever use one message instance. This is achieved by having a cycle in the graph and kickstarting the communication externally by having one of the nodes publish before spinning executor.
+
+### 3. Image Pipeline All In One
+
+Please ensure you have a camera connected to your workstation.
+
+Running `image_pipeline_all_in_one` essentially runs narrow processes of `camera_node`
```

Comment:
What is meant by "narrow processes?"

Reply:
The term `narrow processes` is referenced and paraphrased from the mention of `narrow tasks` under section, **Background**, line 1, in the  official [ROS2 documentation](https://docs.ros.org/en/rolling/Tutorials/Demos/Intra-Process-Communication.html) for `intra_process_demo`.

It likely refers to the how each node's functionality is singular and easy-to-understand.

Eg.

`camera_node` focuses on coverting capture images to ROS2 image messages.
`watermark_node` focuses on overlaying text on the ROS2 image messages.
`image_viewer_node` focuses on displaying the overlayed image messages in a window for users.

Reply:
I think it would enhance comprehension to avoid introducing jargon, even if it is used elsewhere in ROS documentation. In this case, one could simply describe what running image_pipeline_all_in_one does. For example,

image_pipeline_all_in_one consists of 3 nodes, where the first one publishes a unique_ptr<Image> message, the second node subscribes to the topic and republishes the image after modifying it on a new topic, and the third node subscribes to to this last topic.

Reply:
Modified the identified paragraph to the following based on feedback:

> `image_pipeline_all_in_one` consists of **3** nodes, where `camera_node` publishes a `unique_ptr` message onto the topic, `\image`. Subsequently,  `watermark_node` subscribes to the previous, republishes the image after adding a watermark on the topic, `\watermarked_image`, and the final node, `image_view_node`, subscribes to this last topic in order to display it for user.